src/shell/commands/executable.rs

```diff
@@ -55,24 +55,44 @@ impl ShellCommand for ExecutableCommand {
       // avoid deadlock since this is holding onto the pipes
       drop(sub_command);
 
-      tokio::select! {
-        result = child.wait() => match result {
-          Ok(status) => ExecuteResult::Continue(
-            status.code().unwrap_or(1),
-            Vec::new(),
-            Vec::new(),
-          ),
-          Err(err) => {
-            let _ = stderr.write_line(&format!("{}", err));
-            ExecuteResult::Continue(1, Vec::new(), Vec::new())
+      loop {
+        tokio::select! {
+          result = child.wait() => match result {
+            Ok(status) => return ExecuteResult::Continue(
+              status.code().unwrap_or(1),
+              Vec::new(),
+              Vec::new(),
+            ),
+            Err(err) => {
+              let _ = stderr.write_line(&format!("{}", err));
+              return ExecuteResult::Continue(1, Vec::new(), Vec::new())
+            }
+          },
+          signal = context.state.kill_signal().wait_any() => {
+            if let Some(_id) = child.id() {
+              #[cfg(unix)]
+              kill(_id as i32, signal);
+
+              if cfg!(not(unix)) && signal.causes_abort() {
+                let _ = child.kill().await;
+                return ExecuteResult::Continue(signal.aborted_code(), Vec::new(), Vec::new());
+              }
+            }
           }
-        },
-        _ = context.state.token().cancelled() => {
-          let _ = child.kill().await;
-          ExecuteResult::for_cancellation()
         }
       }
     }
     .boxed_local()
   }
 }
+
+#[cfg(unix)]
+pub fn kill(pid: i32, signal: crate::SignalKind) -> Option<()> {
+  use nix::sys::signal::kill as unix_kill;
+  use nix::sys::signal::Signal;
+  use nix::unistd::Pid;
+  let signo: i32 = signal.into();
+  let sig = Signal::try_from(signo).ok()?;
```

Comment:
Should we print something if a signal can't be converted?

Reply:
I think we should just ignore it.